src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -162,7 +162,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mpi_isend(const void *buf, MPI_Aint count
 
     MPIDI_Datatype_check_contig_size(datatype, count, dt_contig, data_sz);
     if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE > -1 &&
-        dt_contig && data_sz > MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE) {
+        dt_contig && data_sz > MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE && rank != MPIR_Process.rank) {
```

Comment:
The code seems incorrect. `rank` is the receiver's rank in comm, but `MPIR_Process.rank` is the local process's rank in COMM_WORLD. I think you should compare `rank != comm->rank`.  Once you change the code, please also run some self-message tests to make sure it actually goes to POSIX.

Reply:
You are right, sorry for this bug. I should have used comm->rank.